jetty-ee9/jetty-ee9-tests/jetty-ee9-test-jersey/src/test/java/org/eclipse/jetty/ee9/jersey/tests/HungBlockingThreadsTest.java

```diff
@@ -0,0 +1,230 @@
+        queuedThreadPool.setDaemon(true);
+        queuedThreadPool.setReservedThreads(1);
+
+        server = new Server(queuedThreadPool);
+
+        ServletContextHandler context = new ServletContextHandler(ServletContextHandler.SESSIONS);
+        context.setContextPath("/");
+
+        ResourceConfig resourceConfig = new ResourceConfig();
+        Resource resource = Resource.builder(PostResource.class).build();
+        Resource.Builder resourceBuilder = Resource.builder(resource);
+        resource.getResourceMethods().forEach(resourceMethod -> resourceBuilder.updateMethod(resourceMethod).managedAsync());
+        resourceConfig.registerResources(resourceBuilder.build());
+        resourceConfig.register(CustomFeature.class);
+
+        context.addServlet(new ServletHolder(new ServletContainer(resourceConfig)), "/*");
+
+        server.setHandler(context);
+
+        ServerConnector connector = new ServerConnector(server);
+        connector.setPort(0);
+        connector.setIdleTimeout(Long.MAX_VALUE);
+        server.addConnector(connector);
+        server.start();
+
+        httpClient = new HttpClient();
+        QueuedThreadPool clientQtp = new QueuedThreadPool(THREAD_COUNT * 2);
+        clientQtp.setName("client");
+        httpClient.setExecutor(clientQtp);
+        httpClient.setMaxConnectionsPerDestination(THREAD_COUNT);
+        httpClient.start();
+
+        executorService = Executors.newCachedThreadPool();
+    }
+
+    @AfterEach
+    public void tearDown()
+    {
+        executorService.shutdownNow();
+        LifeCycle.stop(server);
+        LifeCycle.stop(httpClient);
+    }
+
+    @Test
+    public void test() throws Exception
+    {
+        CountDownLatch latch = new CountDownLatch(THREAD_COUNT);
+        for (int i = 0; i < THREAD_COUNT; i++)
+        {
+            executorService.submit(() ->
+            {
+                httpClient.POST(server.getURI())
+                    .body(new InputStreamRequestContent("text/plain", new HangInputStream(latch)))
+                    .timeout(1, TimeUnit.DAYS)
+                    .send(result -> {});
+            });
+        }
+
+        LOG.debug("awaiting for client to block on all threads");
+        assertTrue(latch.await(15, TimeUnit.SECONDS));
+
+        LOG.debug("Shutting down client");
+        // while hanging in inputstream read, close connections.
+        LifeCycle.stop(httpClient);
+
+        LOG.debug("Waiting for hung threads");
+        await().atMost(5, TimeUnit.SECONDS).untilAsserted(() -> assertThat(threadDump(), not(containsString(SharedBlockingCallback.class.getName()))));
+    }
+
+    private static class HangInputStream extends InputStream
+    {
+        private final CountDownLatch latch;
+        private int count = 0;
+
+        public HangInputStream(CountDownLatch latch)
+        {
+            this.latch = latch;
+        }
+
+        @Override
+        public int read()
+        {
+            if (count == 1_000_000)
```

Comment:
Why 1M?

Reply:
That's not really needed; 1 KB of data upload seems sufficient to reproduce the issue.